aptos/api/aux-ts/src/indexer/index.ts

```diff
@@ -86,188 +87,161 @@ async function publishAmmEvents() {
-    // only process the records within a page that are > cursor
-    // move the cursor forward after processing
-    const moreFills = await market.fills({ start: cursor, limit: 100 });
-    const fillsSince =
-      cursor === 0
-        ? moreFills
-        : _.dropWhile(
-            moreFills,
-            (fill) => fill.sequenceNumber.toNumber() <= cursor
-          );
-    fills = fills.concat(fillsSince.map(convert));
-    cursor =
-      fillsSince[fillsSince.length - 1]?.sequenceNumber.toNumber() ?? cursor;
-    if (fillsSince.length === 0 || fillsSince.length < moreFills.length) {
-      break;
-    }
+  if (!_.isUndefined(cursor)) {
+    keyToCursor[key] = cursor;
+    // publish bars
+    await Promise.all(
+      RESOLUTIONS.map((resolution) => publishBar(resolution, key))
+    );
   }
+}
+
+function publishFills(fills: Order[], key: string) {
+  return Promise.all([
+    fills.forEach(async (fill) => await redisPubSub.publish("TRADE", fill)),
+    fills.forEach(
+      async (fill) => await redisPubSub.publish("LAST_TRADE_PRICE", fill.price)
+    ),
+    redisClient.rPush(
+      `${key}-fills-24h`,
+      fills.map((fill) =>
+        JSON.stringify({
+          time: fill.time,
+          price: fill.price,
+          quantity: fill.quantity,
+        })
+      )
+    ),
+    Promise.all(
+      RESOLUTIONS.map((resolution) => {
+        redisClient.rPush(
+          `${key}-fills-${resolution}`,
+          fills.map((fill) =>
+            JSON.stringify({
+              time: fill.time,
+              price: fill.price,
+              quantity: fill.quantity,
+            })
+          )
+        );
+      })
+    ),
+  ]);
+}
+
+async function publishBar(resolution: Resolution, key: string): Promise<void> {
+  key = `${key}-fills-${resolution}`;
+  const now = Date.now();
+  const delta = now % (resolutionToSeconds(resolution) * 1000);
+  const time = now - delta;
+
+  const len = await redisClient.lLen(key);
+  const rawFills = await redisClient.lPopCount(key, len) ?? [];
+  const fills: Order[] = rawFills
+    .map((s) => JSON.parse(s))
+    .filter((fill: Order) => Number(fill.time) >= time);
 
+  if (_.isEmpty(fills)) {
+    await redisPubSub.publish("BAR", { resolution, time, ohlcv: null });
+    return;
+  }
   const prices = fills.map((fill) => fill.price);
   const bar: Bar = {
-    open: _.first(prices),
-    high: _.max(prices),
-    low: _.max(prices),
-    close: _.last(prices),
-    volume: _(fills)
-      .map((fill) => fill.price * fill.quantity)
-      .sum(),
-    fills,
-    cursor,
+    resolution,
+    time,
+    ohlcv: {
+      open: _.first(prices)!,
+      high: _.max(prices)!,
+      low: _.max(prices)!,
+      close: _.last(prices)!,
+      volume: _(fills)
+        .map((fill) => fill.price * fill.quantity)
+        .sum(),
+    },
   };
-
-  fs.writeFileSync(PATH, JSON.stringify(bar));
-  return bar;
+  await redisPubSub.publish("BAR", bar);
```

Comment:
It seems like you're publishing bars for every resolution once per second. Is trading view able to resolve that into bars of various resolutions? E.g., for 1m resolution does it take the last bar update from every minute as the historical bar?

Reply:
yep, every bar is published with a "time" which is the start time of the interval for trading view to use

we might publish the same bar multiple times (e.g. every second for those 15 seconds) as an updated view

moreover, there's of course a time-gap between when it actually happened and when the producer polls it, so there's kinda the two axis of time processing vs. event time

Reply:
Oh I see, `time` is the start time. Makes sense.